[docs] Add guide for code references (#21755)
dagster-io/dagster: docs/content/guides/dagster/code-references.mdx

---

> title: Linking to assets' code with code references | Dagster Docs

Reviewer: The possessive is a weird stylization in a title. I wonder if we can circumvent it altogether. We already use "asset definition" below anyways.

```suggestion
title: Linking to asset definition code with code references | Dagster Docs
```

---

> title: Code References in Dagster | Dagster Docs

Reviewer 2: ```suggestion
title: Linking to code in Dagster with Code References | Dagster Docs
```

Making this match the page title

---

> # Linking to code in Dagster with Code References

Reviewer 2: Should we mention assets in the title, since this seems to be specific to assets? Are we planning to expand this to other types of definitions?

Author: Good call, updated!

<Note>
  This feature is considered <strong>experimental</strong> and is under active
  development. This guide will be updated as we roll out new features.
</Note>

> This guide explains how to use Code Reference metadata on your Dagster assets to link from the Dagster UI to the source code that produced the asset. In the local development environment, you can use this feature to navigate directly to the code backing an asset in your editor, and in the Dagster Cloud UI, you can link to the source code in your source control repository to see the full history of changes.

Reviewer 2: Thoughts on something like this to break things up a bit?

```suggestion
Using Code Reference metadata on your Dagster assets, you can easily view the asset's source code from the Dagster UI:

- **In local development**, navigate directly to the code backing an asset in your editor
- **In Dagster+**, link to source code in your source control repository to see the full change history for an asset

In this guide, we'll show you how to automatically and manually attach Code References to your Dagster assets.
```

---

## Prerequisites

To complete the steps in this guide, you'll need:

- A Dagster repository with assets that you want to link to code
- Dagster version `1.7.6` or newer

---

## Referencing source code in your assets

> ### Automatically attaching code references to asset definitions

Reviewer 2: Because there's only one instructional section in this guide, I'd say we scrap `Referencing source code in your assets` and making the other two headings into H2s. If we plan to expand this to other definitions, though, maybe it's fine to keep it.

To automatically attach code references to the asset's function definition, you can use the `dagster._core.definitions.metadata.with_source_code_references` utility. Any asset definitions passed to the utility will have their source file attached as metadata.

> ```python file=/guides/dagster/code_references/with_source_code_references.py

Reviewer 2: I suggest we include the name of the file in this example so folks can make the connection between this and what is shown in the screen shot of the UI.

Author: Good call, updated!

from dagster import Definitions, asset
from dagster._core.definitions.metadata import with_source_code_references

@asset
def my_asset(): ...

@asset
def another_asset(): ...

defs = Definitions(
    assets=with_source_code_references(
        [
            my_asset,
            another_asset,
        ]
    )
)
```

> A link to source will then be visible in the Asset Catalog view in the Dagster UI:

Reviewer 2: Isn't this the **Asset details** page? Also, perhaps include the name of the field (**Code location**) so folks can immediately find it in the image

Author: Updated 👍

<center>
  <Image
    alt="Asset catalog view showing link to with_source_code_references.py"
    src="/images/guides/code-references/with_source_code_references.png"
    width={1590}
    height={326}
  />
</center>

### Manually attaching code references to asset definitions

> In some cases, you may want to manually attach code references to your asset definitions. For example, an asset whose definition is spread across multiple Python sources files or is partially defined in YAML may have a more complex source structure.

Reviewer 2: ```suggestion
In some cases, you may want to manually attach code references to your asset definitions. For example, an asset whose definition is spread across multiple Python source files or is partially defined in YAML may have a more complex source structure.
```

For the YAML files - what does this refer to? Something like dbt models?

Author: Updated to more clearly explain the example case.

> To manually attach code references to an asset definition, you can manually attach an instance of `dagster._core.definitions.metadata.CodeReferencesMetadataSet`. You can then choose to augment these manual references with `with_source_code_references`:

Reviewer 2: ```suggestion
To manually attach code references to an asset definition, use an instance of `dagster._core.definitions.metadata.CodeReferencesMetadataSet`. You can then choose to augment these manual references with `with_source_code_references`:
```

Removing some redundancy

> ```python file=/guides/dagster/code_references/manual_references.py

Reviewer 2: Same suggestion here about including the name of the file somewhere

---
title: Linking to assets' code with code references | Dagster Docs
---

# Linking to assets' code with code references

<Note>
  This feature is considered <strong>experimental</strong> and is under active
  development. This guide will be updated as we roll out new features.
</Note>

> Attaching code reference metadata to your Dagster assets allows you to easily view those assets' source code from the Dagster UI:

Reviewer 3: ```suggestion
Attaching code reference metadata to your Dagster asset definitions allows you to easily view those assets' source code from the Dagster UI:
```

- **In local development**, navigate directly to the code backing an asset in your editor

> - **In Dagster+**, link to source code in your source control repository to see the full change history for an asset

Reviewer 3: This isn't just for Dagster+, right? They can also do this in their OSS deployment.

> In this guide, we'll show you how to automatically and manually attach Code References to your Dagster assets.

Reviewer 2: You've got `Code References` here, and everywhere else it's `code references`. If this isn't a proprietary feature, lets go with `code references` to be consistent.

---

## Prerequisites

To complete the steps in this guide, you'll need:

> - A Dagster repository with assets that you want to link to code

Reviewer 3: "repository" is mostly a deprecated term. Maybe just "a set of Dagster asset definitions"?

- Dagster version `1.7.6` or newer

---

## Automatically attaching code references to asset definitions

To automatically attach code references to the asset's function definition, you can use the `dagster._core.definitions.metadata.with_source_code_references` utility. Any asset definitions passed to the utility will have their source file attached as metadata.

```python file=/guides/dagster/code_references/with_source_code_references.py
from dagster import Definitions, asset

> from dagster._core.definitions.metadata import with_source_code_references

Reviewer 3: We should expose this publicly if we're going to include it in a guide

@asset
def my_asset(): ...

@asset
def another_asset(): ...

defs = Definitions(
    assets=with_source_code_references(
        [
            my_asset,

>             another_asset,

Reviewer 3: taking away the comma at the end of this line would allow this snippet to compose more

        ]
    )
)
```

> A link to the asset's source in `with_source_code_references.py` will then be visible in the Asset Catalog view in the Dagster UI:

Reviewer 2: ```suggestion
A link to the asset's source in `with_source_code_references.py` will then be visible in the **Asset Catalog** view in the Dagster UI:
```

pls format UI components (page names, buttons, copy, etc) in **bold**

<center>
  <Image
    alt="Asset catalog view showing link to with_source_code_references.py"
    src="/images/guides/code-references/with_source_code_references.png"
    width={1590}
    height={326}
  />
</center>

---

## Manually attaching code references to asset definitions

In some cases, you may want to manually attach code references to your asset definitions. Some assets may have a more complex source structure, such as an asset whose definition is spread across multiple Python source files or an asset which is partially defined with a `.sql` model file.

> To manually attach code references to an asset definition, use an instance of `dagster._core.definitions.metadata.CodeReferencesMetadataSet`. You can then choose to augment these manual references with `with_source_code_references`:

Reviewer 3: I think we're not ready yet to talk about `MetadataSet`s publicly.  We should just recommend
```python
metadata={
    "dagster/code_references": CodeReferencesMetadataValue(

...
```

similar to what we do in the column lineage guide: https://docs.dagster.io/concepts/metadata-tags/asset-metadata/column-level-lineage#enabling-column-level-lineage

```python file=/guides/dagster/code_references/manual_references.py
import os

from dagster import Definitions, asset
from dagster._core.definitions.metadata import (
    CodeReferencesMetadataSet,
    CodeReferencesMetadataValue,
    LocalFileCodeReference,
    with_source_code_references,
)

@asset(
    metadata={
        **CodeReferencesMetadataSet(
            code_references=CodeReferencesMetadataValue(
                code_references=[
                    LocalFileCodeReference(
                        file_path=os.path.join(
                            os.path.dirname(__file__), "source.yaml"
                        ),

>                         line_number=1,

Reviewer 3: This makes me think we should make `line_number` optional

                        label="Model YAML",
                    )
                ]
            )
        )
    }
)
def my_asset_modeled_in_yaml(): ...

defs = Definitions(assets=with_source_code_references([my_asset_modeled_in_yaml]))
```

> Each of the code references to `manual_references.py` will be visible in the asset details view in the Dagster UI:

Reviewer 2: ```suggestion
Each of the code references to `manual_references.py` will be visible in the **Asset details** page in the Dagster UI:
```

Same here, nitpicking about formatting UI stuff ⛏️
